network/libp2p_impl.go

```diff
@@ -10,36 +10,96 @@ import (
 	"github.com/libp2p/go-libp2p-core/host"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
 	"github.com/libp2p/go-msgio"
 	ma "github.com/multiformats/go-multiaddr"
 
 	gsmsg "github.com/ipfs/go-graphsync/message"
+	gsmsgv1 "github.com/ipfs/go-graphsync/message/v1"
+	gsmsgv2 "github.com/ipfs/go-graphsync/message/v2"
 )
 
 var log = logging.Logger("graphsync_network")
 
 var sendMessageTimeout = time.Minute * 10
 
+// Option is an option for configuring the libp2p storage market network
+type Option func(*libp2pGraphSyncNetwork)
+
+// DataTransferProtocols OVERWRITES the default libp2p protocols we use for
+// graphsync with the specified protocols
+func GraphsyncProtocols(protocols []protocol.ID) Option {
+	return func(gsnet *libp2pGraphSyncNetwork) {
+		gsnet.setProtocols(protocols)
+	}
+}
+
 // NewFromLibp2pHost returns a GraphSyncNetwork supported by underlying Libp2p host.
-func NewFromLibp2pHost(host host.Host) GraphSyncNetwork {
+func NewFromLibp2pHost(host host.Host, options ...Option) GraphSyncNetwork {
+	messageHandlerSelector := messageHandlerSelector{
+		v1MessageHandler: gsmsgv1.NewMessageHandler(),
+		v2MessageHandler: gsmsgv2.NewMessageHandler(),
+	}
 	graphSyncNetwork := libp2pGraphSyncNetwork{
-		host: host,
+		host:                   host,
+		messageHandlerSelector: &messageHandlerSelector,
+		protocols:              []protocol.ID{ProtocolGraphsync_2_0_0, ProtocolGraphsync_1_0_0},
+	}
+
+	for _, option := range options {
+		option(&graphSyncNetwork)
 	}
 
 	return &graphSyncNetwork
 }
 
+// a message.MessageHandler that simply returns an error for any of the calls, allows
+// us to simplify erroring on bad protocol within the messageHandlerSelector#Select()
+// call so we only have one place to be strict about allowed versions
+type messageHandlerErrorer struct {
+	err error
+}
+
+func (mhe messageHandlerErrorer) FromNet(peer.ID, io.Reader) (gsmsg.GraphSyncMessage, error) {
+	return gsmsg.GraphSyncMessage{}, mhe.err
+}
+func (mhe messageHandlerErrorer) FromMsgReader(peer.ID, msgio.Reader) (gsmsg.GraphSyncMessage, error) {
+	return gsmsg.GraphSyncMessage{}, mhe.err
+}
+func (mhe messageHandlerErrorer) ToNet(peer.ID, gsmsg.GraphSyncMessage, io.Writer) error {
+	return mhe.err
+}
+
+type messageHandlerSelector struct {
+	v1MessageHandler gsmsg.MessageHandler
+	v2MessageHandler gsmsg.MessageHandler
+}
+
+func (smh messageHandlerSelector) Select(protocol protocol.ID) gsmsg.MessageHandler {
+	switch protocol {
+	case ProtocolGraphsync_1_0_0:
+		return smh.v1MessageHandler
+	case ProtocolGraphsync_2_0_0:
+		return smh.v2MessageHandler
+	default:
+		return messageHandlerErrorer{fmt.Errorf("unrecognized protocol version: %s", protocol)}
```

Comment:
I'd quite like to test this but don't know how! I'd need to set up a host, then connect to it with a stream with an entirely different protocol stream. Hints would be good.